docs: Add nico-mcp operator guide (#3498)
NVIDIA/bare-metal-manager-core: docs/operations/nico-mcp.md

---
title: "NICo MCP Server"
description: "Deploy and operate the NICo MCP server, including configuration precedence and security requirements."
---

# NICo MCP Server

`nico-mcp` exposes the read-only NICo REST API surface as Model Context
Protocol (MCP) tools over streamable HTTP. It generates tools for the `GET`
operations in the embedded OpenAPI specification and does not expose REST
operations that create, update, or delete resources.

The server is stateless. It does not authenticate or authorize callers. NICo
REST validates the bearer token and enforces organization and role permissions
for every forwarded request.

## Supported Deployment Topologies

### Production: ClusterIP Alongside NICo REST

The supported production topology deploys `nico-mcp` in Kubernetes alongside
NICo REST with the Helm chart at `helm/rest/nico-mcp`. By default, the chart
creates a `ClusterIP` Service on port `8080` and serves MCP requests at `/mcp`.

Configure one fixed NICo REST base URL at startup. An in-cluster Service URL is

> the normal choice:

Reviewer: ```suggestion
the typical choice:
```

Author: Thanks — I’ll take the sentence, list, table-formatting, and capitalization edits. I also checked the proposed precedence wording against current `main`; it matches the behavior merged in #3499.

```yaml
config:
  baseURL: http://nico-rest-api:8388
  apiName: nico
```

The MCP listener is plaintext HTTP. Place an upstream service in front of it
that provides all of the following controls:

- authenticates callers and passes each caller's bearer token to `nico-mcp`;
- terminates TLS for client connections;
- routes MCP traffic to the `nico-mcp` ClusterIP Service; and

> - restricts direct access to the MCP Service from untrusted networks.

Reviewer: ```suggestion
- Authenticates callers and passes each caller's bearer token to `nico-mcp`
- Terminates TLS for client connections
- Routes MCP traffic to the `nico-mcp` ClusterIP Service
- Restricts direct access to the MCP Service from untrusted networks
```

Author: Thanks — I’ll take the sentence, list, table-formatting, and capitalization edits. I also checked the proposed precedence wording against current `main`; it matches the behavior merged in #3499.

The chart does not deploy these controls. Use the ingress, gateway, proxy, and
network-policy components approved for your environment.

### Standalone Development and Testing

For development and testing, build and run the binary directly against a
reachable NICo REST endpoint:

```bash
cd rest-api
make nico-mcp
nico-mcp \
  --listen 127.0.0.1:8080 \
  --path /mcp \
  --base-url https://nico.example.com \
  --org example-org
```

Keep the listener on localhost or a trusted network. If it must be reachable
from an untrusted network, put it behind the same authentication, TLS, routing,
and network controls required for production.

## Install the Helm Chart

Before installing the chart, make sure that:

- the `nico-mcp` image is available to the cluster;
- NICo REST is reachable from the target namespace; and

> - the upstream authentication and TLS front is configured.

Reviewer: ```suggestion
- The `nico-mcp` image is available to the cluster
- NICo REST is reachable from the target namespace
- The upstream authentication and TLS frontend is configured
```

Author: Thanks — I’ll take the sentence, list, table-formatting, and capitalization edits. I also checked the proposed precedence wording against current `main`; it matches the behavior merged in #3499.

Create a values file for the deployment:

```yaml
global:
  image:
    repository: registry.example.com/nico
    tag: "2.0.0"
    pullPolicy: IfNotPresent
  imagePullSecrets:
    - name: image-pull-secret

config:
  baseURL: http://nico-rest-api:8388
  org: example-org
  apiName: nico
```

Install the chart in the NICo REST namespace:

```bash
helm upgrade --install nico-mcp helm/rest/nico-mcp \
  --namespace nico-rest \
  --create-namespace \
  --values nico-mcp-values.yaml
```

The rendered image is
`<global.image.repository>/<image.name>:<global.image.tag>`. The chart does not
create `global.imagePullSecrets`; create the referenced Secret before the
Deployment starts when the registry requires authentication.

### Chart Values

| Value | Default | Purpose |

> | --- | --- | --- |

Reviewer: Make it a bit easier to read... (helps a lot with more columns)

```suggestion
| Value | Default | Purpose |
| ----- | ------- | ------- |
```

Author: Thanks — I’ll take the sentence, list, table-formatting, and capitalization edits. I also checked the proposed precedence wording against current `main`; it matches the behavior merged in #3499.

| `global.image.repository` | empty | Image registry and repository prefix. Required for deployment. |
| `global.image.tag` | empty | Image tag. Required for deployment. |
| `global.image.pullPolicy` | `IfNotPresent` | Kubernetes image pull policy. |

> | `global.imagePullSecrets` | `image-pull-secret` | Existing image pull secrets attached to the pod. |

Reviewer 2: NIT (non-blocking): the actual default is a list of name objects, and the value's shape matters — `global.imagePullSecrets: my-secret` would fail to render. Since the other rows show paste-able literals, consider:

```suggestion
| `global.imagePullSecrets` | `[{name: image-pull-secret}]` | Existing image pull secrets attached to the pod (list of `name:` objects). |
```

> | `global.imagePullSecrets` | `image-pull-secret` | Existing image pull Secrets attached to the Pod. |

Reviewer: I think we could capitalize fewer words here :)

```suggestion
| `global.imagePullSecrets` | `image-pull-secret` | Existing image pull secrets attached to the pod. |
```

Author: Thanks — I’ll take the sentence, list, table-formatting, and capitalization edits. I also checked the proposed precedence wording against current `main`; it matches the behavior merged in #3499.

| `replicaCount` | `1` | Number of MCP server replicas. |
| `nameOverride` | empty | Override for the Deployment and Service name. |
| `namespaceOverride` | empty | Namespace override; otherwise the Helm release namespace is used. |
| `image.name` | `nico-mcp` | Image name appended to the repository. |
| `service.type` | `ClusterIP` | Kubernetes Service type. Keep `ClusterIP` for production. |
| `service.port` | `8080` | Listener and Service port. |
| `config.path` | `/mcp` | HTTP path for the MCP handler. |

> | `config.shutdownTimeout` | `10s` | Graceful shutdown timeout. |

Reviewer: Does `10s` need to be double-quoted as `"10s"` here? I know times are done that way in some other spots in the docs

Author: `10s` does not need YAML quotes here. I verified the chart accepts it as a string and `helm template` renders the container argument as `"10s"`. I’ll keep the documented default aligned with the chart source.

| `config.baseURL` | empty | Fixed default NICo REST URL. Set this in production. |
| `config.org` | empty | Default organization for tool calls. |
| `config.apiName` | empty | REST API path segment; the binary defaults to `nico`. |
| `config.debug` | `false` | Log outbound NICo REST HTTP requests and responses. |
| `resources.requests` | `50m` CPU, `64Mi` memory | Pod resource requests. |

> | `resources.limits` | `250m` CPU, `256Mi` memory | Pod resource limits. |

Reviewer: Same double-quote question here

Author: `10s` does not need YAML quotes here. I verified the chart accepts it as a string and `helm template` renders the container argument as `"10s"`. I’ll keep the documented default aligned with the chart source.

Bearer tokens are intentionally not exposed as chart values. Shared production
deployments should use the authenticated caller's token instead of a static
service token.

## Runtime Configuration

> The standalone binary accepts these startup settings. When running the binary

Reviewer: ```suggestion
The standalone binary accepts the following startup settings. When running the binary
```

Author: Thanks — I’ll take the sentence, list, table-formatting, and capitalization edits. I also checked the proposed precedence wording against current `main`; it matches the behavior merged in #3499.

directly, a command-line flag takes its value from the corresponding environment
variable when the flag is omitted.

| Flag | Environment variable | Default | Helm value |

> | --- | --- | --- | --- |

Reviewer: ```suggestion
| Flag | Environment variable | Default | Helm value |
| ---- | -------------------- | ------- | ---------- |
```

Author: Thanks — I’ll take the sentence, list, table-formatting, and capitalization edits. I also checked the proposed precedence wording against current `main`; it matches the behavior merged in #3499.

| `--listen` | `NICO_MCP_LISTEN` | `:8080` | `service.port` sets the port |
| `--path` | `NICO_MCP_PATH` | `/mcp` | `config.path` |
| `--shutdown-timeout` | `NICO_MCP_SHUTDOWN_TIMEOUT` | `10s` | `config.shutdownTimeout` |
| `--base-url` | `NICO_BASE_URL` | none | `config.baseURL` |
| `--org` | `NICO_ORG` | none | `config.org` |
| `--api-name` | `NICO_API_NAME` | `nico` | `config.apiName` |

> | `--token` | `NICO_TOKEN` | none | not exposed |

Reviewer: ```suggestion
| `--token` | `NICO_TOKEN` | none | Not exposed |
```

Author: Thanks — I’ll take the sentence, list, table-formatting, and capitalization edits. I also checked the proposed precedence wording against current `main`; it matches the behavior merged in #3499.

| `--debug` | none | `false` | `config.debug` |

The Helm chart explicitly renders `--listen`, `--path`, and
`--shutdown-timeout` from `service.port`, `config.path`, and
`config.shutdownTimeout`. Those Helm values take precedence over the
corresponding `NICO_MCP_*` environment variables.

`nico-mcp` does not read `~/.nico/config.yaml`.

Every tool accepts optional `base_url`, `org`, `api_name`, and `token`
arguments. Non-empty values are resolved for each call as follows:

| Setting | Precedence, highest first |

> | --- | --- |

Reviewer: ```suggestion
| Setting | Precedence, highest first |
| ------- | ------------------------- |
```

Author: Thanks — I’ll take the sentence, list, table-formatting, and capitalization edits. I also checked the proposed precedence wording against current `main`; it matches the behavior merged in #3499.

> | `base_url` | Fixed startup `--base-url` or `NICO_BASE_URL`; without a startup value, the tool argument |

Reviewer 2: NIT (non-blocking): this row lists the startup value first while the other three rows lead with the tool argument, so a reader scanning the table can miss that when both are set they must be *equal* (mismatch is rejected) rather than one silently winning. The paragraph below does clarify it — maybe just add "(must match; mismatch rejected)" to the row.

| `org` | tool argument, then startup `--org` or `NICO_ORG` |
| `api_name` | tool argument, then startup `--api-name` or `NICO_API_NAME`, then `nico` |

> | `token` | tool argument, inbound `Authorization: Bearer` header, then startup `--token` or `NICO_TOKEN` |

Reviewer: ```suggestion
| `base_url` | Fixed startup `--base-url` or `NICO_BASE_URL`; without a startup value, the tool argument |
| `org` | Tool argument, then startup `--org` or `NICO_ORG` |
| `api_name` | Tool argument, then startup `--api-name` or `NICO_API_NAME`, then `nico` |
| `token` | Tool argument, inbound `Authorization: Bearer` header, then startup `--token` or `NICO_TOKEN` |
```

Author: I verified this against current `main` after #3499. With a configured startup base URL, a call may omit `base_url` or repeat the same destination under the server’s URL identity comparison; a different destination is rejected before the outbound client is constructed. Without a configured startup URL, a caller-selected destination is allowed with an explicit per-call token or with no resolved token. Inbound and startup credentials are rejected for that caller-selected destination. I’ll rebase the guide onto `main` and preserve that distinction.

> When a startup base URL is configured, a tool argument may repeat that

Reviewer: ```suggestion
When a startup base URL is configured, a tool argument can repeat that
```

Author: I verified this against current `main` after #3499. With a configured startup base URL, a call may omit `base_url` or repeat the same destination under the server’s URL identity comparison; a different destination is rejected before the outbound client is constructed. Without a configured startup URL, a caller-selected destination is allowed with an explicit per-call token or with no resolved token. Inbound and startup credentials are rejected for that caller-selected destination. I’ll rebase the guide onto `main` and preserve that distinction.

destination, with or without a trailing slash, but it cannot select a different
one. When no startup base URL is configured, a tool call can supply `base_url`
only with an explicit per-call `token` or with no resolved token. The server
does not combine a caller-chosen destination with an inherited inbound or
startup token.

`base_url` and `org` must resolve to non-empty values. A token can be empty,
but NICo REST normally rejects the resulting unauthenticated request.

## Security Requirements

Treat the REST destination and the bearer token as one trust decision:

- Set a fixed startup base URL for every production deployment.
- Use caller-specific bearer tokens on shared endpoints. NICo REST remains the
  authorization enforcement point.
- A configured startup base URL pins the REST destination. A different
  per-call `base_url` is rejected, including when the call supplies an explicit
  token.
- Without a startup base URL, a per-call `base_url` is allowed only when the

>   call also supplies its own explicit `token`, or when the request resolves to

Reviewer: ```suggestion
  call also supplies its own explicit token, or when the request resolves to
```

Author: I verified this against current `main` after #3499. With a configured startup base URL, a call may omit `base_url` or repeat the same destination under the server’s URL identity comparison; a different destination is rejected before the outbound client is constructed. Without a configured startup URL, a caller-selected destination is allowed with an explicit per-call token or with no resolved token. Inbound and startup credentials are rejected for that caller-selected destination. I’ll rebase the guide onto `main` and preserve that distinction.
